### ISSUE TITLE
PennyLane does not allow multiple state preparations at the beginning of a circuit

### ISSUE THREAD

**Issue by sassan72** [PennyLaneAI/pennylane#1140]
i want to design a swap classifier using amplitude encoding method Pennylane. But i got this error:

```
pennylane._device.DeviceError: Operation QubitStateVector cannot be used after other Operations have already been applied on a default.qubit.autograd device.
```

```python
@qml.qnode(dev)
def swap_test(x1, x2):

    # load the two inputs into two different registers
    qml.templates.embeddings.AmplitudeEmbedding(x1, wires=[1, 2, 3, 4, 5])
    qml.templates.embeddings.AmplitudeEmbedding(x2, wires=[6, 7, 8, 9, 10])
   
    qml.Hadamard(wires=0)
    
    qml.CSWAP(wires=[0, 1, 6])
    qml.CSWAP(wires=[0, 2, 7])
    qml.CSWAP(wires=[0, 3, 8])
    qml.CSWAP(wires=[0, 4, 9])
    qml.CSWAP(wires=[0, 5, 10])

    qml.Hadamard(wires=0)

    return qml.expval(qml.PauliZ(0))
```
as you can see the second `qml.templates.embeddings.AmplitudeEmbedding(x2, wires=[6, 7, 8, 9, 10])` is at the first of circuit. It should work without error. But it did’t.

**Comment by mariaschuld** [comment]
Ah, what's happening here is the following:

`AmplitudeEmbedding` calls `QubitStatePreparation`, which in turn "hacks" the initial state in simulators. When `QubitStatePreparation` is not supported on a device, the hack is replaced by a "proper" decomposition into a circuit, but it involves an extremely long circuit with lots of classical pre-processing. The reason for this is that this embedding implements an arbitrary state preparation routine, whose runtime is necessarily exponential in the number of qubits. 

The problem you are encountering is that this "hack", i.e. to replace the device's state with your feature vector to significantly speed up the simulations, cannot be applied twice (it would simply overwrite the state). 

We could think of adding a flag to `AmplitudeEmbedding`  that allows a user to decide whether or not a decomposition is used? 

A short-term fix is for you to use [MottonnenStatePreparation](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.templates.state_preparations.MottonenStatePreparation.html), which is the decomposition that PennyLane would use anyways. 

But, and as a reason for why we did not make this use case a priority, what you are doing in this kernel is basically to compute the inner product of x1 and x2, but using an awfully  complicated quantum circuit. Unless you have a very specific reason to do this, it may not be the best idea in the first place?

**Comment by sassan72** [comment]
thanks for your clarification.

**Comment by mariaschuld** [comment]
Oh wait, I misread your main point, you are applying `AmplitudeEmbedding` to different wires! 

I think in principle we could support the efficient, simulator-compatible implementation of `QubitStateVec` to allow for this, but it would require a little work. 

Will the `MottonnenStatePreparation` trick work for you in the meantime?

**Comment by sassan72** [comment]
Yes. I just wanted to investigate different encoding methods.

**Comment by antalszava** [comment]
Opened a new issue https://github.com/PennyLaneAI/pennylane/issues/1150 for adding the feature to allow using multiple `QubitStateVector` operations if acting on different wires. As this is more so a new feature we'd like to add to PennyLane, rather than a malfunction in how PennyLane works at the moment, removing the bug label.

Perhaps another near-term solution could be to concatenate the input features (`x = np.concatenate((x1,x2))`) and calling `qml.templates.embeddings.AmplitudeEmbedding(x, wires=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10])`.

**Comment by josh146** [comment]
This is now solved via #1933 🙂
